.githooks/pre-commit

```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env bash
+
+# Find staged Rust files
+STAGED_FILES=$(git diff --cached --name-only --diff-filter=ACM | grep '\.rs$')
+
+if [[ -z "$STAGED_FILES" ]]; then
+  # No Rust files staged, continue commit
+  exit 0
+fi
+
+# Run clippy on the changed files:
+echo "Running cargo clippy..."
+cargo clippy --workspace --no-deps -- -D warnings
+
+# Check if cargo clippy found any warnings
+if [[ $? -ne 0 ]]; then
+  echo "❌ Commit aborted. Fix the clippy warnings above, stage the changes, and commit again."
+  exit 1
+fi
+
+# Run cargo fmt on staged files
+echo "Running cargo fmt on staged Rust files..."
+
+# Format each staged Rust file
+cargo fmt --quiet -- $STAGED_FILES
+
+# Check if cargo fmt modified any staged files
+CHANGED_FILES=$(git diff --name-only --diff-filter=M $STAGED_FILES)
+
+if [[ -n "$CHANGED_FILES" ]]; then
+  echo "🚨 Cargo fmt modified files:"
+  echo "$CHANGED_FILES"
+  echo ""
+  echo "❌ Commit aborted. Review the formatting changes above, stage them, and commit again."
+  exit 1
+fi
+
+# Everything looks good
+exit 0
```

Comment:
It's rigid to disallow commits for any reason. There is a Github workflow to review, if you'd like, that will prevent a merge.

Reply:
I am used to preventing commits as early as possible. This saves some quota and time spent on CI. So, these are cheap to catch errors on local machine, but expensive on CI. So, review time (at Github) is spent on more important things than formatting for example. 

And a developer when they want to bypass commit hooks, for any emergency they can just do `git commit ...-n...`. The `-n` bypasses commit hooks.

Just something to think about it. But of course, the whole team needs to like and agree on that. :-)